src/System.Collections.Concurrent/tests/Performance/Perf.ConcurrentBag.cs

```diff
@@ -0,0 +1,192 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Xunit.Performance;
+using Xunit;
+
+namespace System.Collections.Concurrent.Tests
+{
+    public class Perf_ConcurrentBag
+    {
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        public void Ctor()
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long innerIters = Benchmark.InnerIterationCount / 10;
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < innerIters; i++)
+                    {
+                        new ConcurrentBag<int>(); // 1
+                        new ConcurrentBag<int>(); // 2
+                        new ConcurrentBag<int>(); // 3
+                        new ConcurrentBag<int>(); // 4
+                        new ConcurrentBag<int>(); // 5
+                        new ConcurrentBag<int>(); // 6
+                        new ConcurrentBag<int>(); // 7
+                        new ConcurrentBag<int>(); // 8
+                        new ConcurrentBag<int>(); // 9
+                        new ConcurrentBag<int>(); // 10
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 10_000_000), MeasureGCAllocations]
+        public void Add()
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long size = Benchmark.InnerIterationCount;
+                var q = new ConcurrentBag<int>();
+
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        q.Add(i);
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void TryTake(bool exists)
+        {
+            foreach (BenchmarkIteration iteration in Benchmark.Iterations)
+            {
+                long size = Benchmark.InnerIterationCount;
+                var b = new ConcurrentBag<int>();
+
+                if (exists)
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        b.Add(i);
+                    }
+                }
+
+                using (iteration.StartMeasurement())
+                {
+                    for (int i = 0; i < size; i++)
+                    {
+                        b.TryTake(out int result);
+                    }
+                }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = 1_000_000), MeasureGCAllocations]
+        public void AddTake()
```

Comment:
I'm curious - did you see noticeable differences in perf between the sum of `Add` + `TryTake` vs 
 just `AddTake`?

Reply:
Yes, as Add/Take repeated reuses the same array slot over and over, whereas Add lots then TryTake lots grows the array to fit all the elements.